Improve math validation

hsorby opened the pull request:
Here we are shifting to using libXml2 to handle the namespaces of elements.  This should work in a larger variety of cases than what is currently handled.

Also, a change is made from the previous validation where the math strings need to self-sufficient for them to validate with no errors.

Addresses #424.

hsorby committed:
Adding initial test for more robust XML namespace handling.

hsorby committed:
Use libeXml2 to do the namespace detection of elements in the document.

hsorby committed:
Implement a libXml2 implementation of cellml units checking for math sub-document elements.

hsorby committed:
Tidy up debugging mess.  Fix error with adding namespace definition to math node.  Get coverage back to 100% after latest changes.

hsorby committed:
Clean up parser tests to separate out the different parts into individual tests.

hsorby committed:
Apply clang format changes to codebase.

hsorby committed:
Merge branch 'develop' into maths_validation_i424

hsorby committed:
Remove debugging code from src/xmlnode.cpp.

hsorby committed:
Remove all #include <iostream> statements that are not required.

hsorby commented on src/parser.cpp:
```diff
@@ -232,6 +232,55 @@ void Parser::ParserImpl::updateModel(const ModelPtr &model, const std::string &i
     loadModel(model, input);
 }
 
+XmlNamespaceMap attributeNamespaces(const XmlNodePtr &node)
```
Could move this function to src/xmlnode.cpp.

nickerso replied:
I would think it would be better suited to `xmlnode.cpp` but not particularly fussed :)

hsorby commented on src/parser.cpp:
```diff
@@ -232,6 +232,55 @@ void Parser::ParserImpl::updateModel(const ModelPtr &model, const std::string &i
+    return namespaceMap;
+}
+
+XmlNamespaceMap determineMissingNamespaces(const XmlNamespaceMap &namespaceMap1, const XmlNamespaceMap &namespaceMap2)
```
Could move this function to src/xmlnode.cpp.

hsorby commented on src/parser.cpp:
```diff
@@ -232,6 +232,55 @@ void Parser::ParserImpl::updateModel(const ModelPtr &model, const std::string &i
+    return undefinedNamespaces;
+}
+
+XmlNamespaceMap traverseTreeForUndefinedNamespaces(const XmlNodePtr &node)
```
Could move this function to src/xmlnode.cpp.

hsorby commented on src/parser.cpp:
```diff
@@ -398,8 +447,17 @@ void Parser::ParserImpl::loadComponent(const ComponentPtr &component, const XmlN
+            XmlNamespaceMap::const_iterator it;
+            for (it = undefinedNamespaces.begin(); it != undefinedNamespaces.end(); ++it) {
+                childNode->addNamespaceDefinition(it->second, it->first);
+            }
```
This is a big change from the previous state.  Now we look through the math tree and determine which namespaces are used and compares them to what is defined.  The missing namespaces are then added to the root node for the math.

hsorby commented on src/validator.cpp:
```diff
@@ -902,11 +884,12 @@ void Validator::ValidatorImpl::validateMath(const std::string &input, const Comp
-    // is not ideal, libxml does not appear to have a better way to remove a namespace declaration from the tree.
+
+    // Remove the cellml namespace definition.
+    mathNode->removeNamespaceDefinition(CELLML_2_0_NS);
```
Here we are doing a proper libXml2 removal of the CellML namespace from the subdocument.

hsorby commented on tests/validator/validator.cpp:
```diff
@@ -1112,7 +1127,7 @@ TEST(Validator, validMathCnElementsMissingCellMLNamespace)
     m->addComponent(c);
 
     v.validateModel(m);
-    EXPECT_EQ(size_t(0), v.errorCount());
+    EXPECT_EQ_ERRORS(expectedErrors, v);
```
This is an important change from previous validation.  Where there were no errors now there are a lot.  This is because the CellML namespace has not been bought through from the enclosing model.

hsorby committed:
Make xmlnode and xmlattribute methods const where possible.

hsorby committed:
Add documentation to removeNamespaceDefinition().

hsorby committed:
Add test for libXml2 to try and determine the capabilities of the library.

hsorby committed:
Improve test for the capabilities of libXml2.

hsorby committed:
Remove multiple cellml namespace attributes from a single element.

hsorby committed:
Remove test for determining libXml2 capabilities.

hsorby committed:
Remove inclusion of CheckCXXSourceRuns from environmentchecks.cmake.

hsorby committed:
Remove test_libxml2.cpp.

hsorby committed:
Remove setting of file compile defintions on tests/validator/validator.cpp.

hsorby committed:
Add test on multiple cellml namespaces.

hsorby committed:
Add simple debugging helper class for printing to stdout.

hsorby committed:
Allow overloading of operators for dbg class.

hsorby committed:
Add method hasNamespaceDefinition to xmlnode class.  Remove CellML 2.0 namespace definitions from any math element.

hsorby committed:
Add documentation to hasNamespaceDefinition method.

hsorby committed:
Apply clang format fix to source code.

hsorby commented:
This is ready for review.

kerimoyle commented:
I've just started looking and (selfishly ...) went straight to the model I wanted to use to see if it could be read.  (This is what I pared down to give the earlier test which segfaulted on my machine).  
The model is the Hodgkin Huxley 1952 from PMR, run through your wee script to translate from 1 to 2, and now using this branch I get a validator error (just switching out your final validation test file name): 

```
Math in component 'membrane' contains 'time' as a bvar ci element but it is already a variable name., 
Math in component 'sodium_channel_m_gate' contains 'time' as a bvar ci element but it is already a variable name., 
W3C MathML DTD error: No declaration for attribute cellml of element math., 
Math in component 'sodium_channel_h_gate' contains 'time' as a bvar ci element but it is already a variable name., 
Math in component 'potassium_channel_n_gate' contains 'time' as a bvar ci element but it is already a variable name.,
```

... I'm not sure whether the glitch is in the file, the translation, or the validation?  The middle error is a new one, but the others I think have been found before by @agarny ?

I'll have a look at the actual code now ... :)

kerimoyle commented on src/xmlnode.h:
```diff
@@ -57,9 +63,57 @@ class XmlNode
+     * and prefix.
+     *
+     * @param uri The @c std::string representation of the XML namespace URI.
+     * @param prefix The @c std::string representatin of the XML namespace prefix.
```
@param prefix The @c std::string representatin of the XML namespace prefix
check spelling ...

hsorby replied:
If you see typos in documentation or comments just fix them.

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -1764,3 +1771,47 @@ TEST(Validator, unitComplexCycle)
+{
+    const std::string math =
+        "<math xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+        "  <ci cellml:units=\"dimensionless\" cellml:units=\"second\">B</ci>\n"
```
I'd thought that we didn't define units on the `ci` elements, only the `cn` ones?

hsorby replied:
It's not defined by the spec. I think.  The spec. doesn't say that you can't do that and somebody is free to write that sort of thing whether it is a validation error is at this point undefined.  I wouldn't mind someone pointing to the section in the spec. that refutes this though.

kerimoyle replied:
In other sections of the validator/parser we do not permit random stuff to be present in the cellml ... not sure why this would be any different?  It's not (currently) explicitly forbidden, but if it is allowed, then we need to read it and validate it against the units defined for that variable elsewhere in the code.  I don't like the idea of having two possibly conflicting definitions permitted, and the user being unaware which is actually being used.

nickerso replied:
CellML 2.0 explicitly specifies that it is invalid to put random stuff in random places. For MathML, we're only using the W3C MathML 2.0 DTD to validate the math strings, so if it is not picked up by that then its not something we can check for. Having said that, once we get "warning", you'd expect this specific warning would be a potential model debugging check that could be useful to avoid that potential misunderstanding about the duplication of a variable's units.

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -1764,3 +1771,47 @@ TEST(Validator, unitComplexCycle)
+
+    const std::vector<std::string> expectedErrors = {
+        "Attribute cellml:units redefined.",
+        "Could not get a valid XML root node from the math on component 'componentName'.",
```
Also don't understand how this error message correlates with the problem of duplicated namespace - not sure how that affects the root node?

hsorby replied:
I think libXml2 takes exception to this really bad xml and gives you nothing back.

kerimoyle replied:
Could we change the error to be something more general then?  "This was so badly formed I didn't know where to start parsing" instead of pointing the user to an error which doesn't make sense?

kerimoyle commented on src/validator.cpp:
```diff
@@ -980,34 +983,23 @@ void Validator::ValidatorImpl::validateAndCleanMathCiCnNodes(XmlNodePtr &node, c
+        bool checkUnitsIsInModel = false;
+        // Check that cellml:units has been set if appropriate for type.
+        if (ciType && unitsAttribute != nullptr) {
+            checkUnitsIsInModel = validateCiCnUnits(component, unitsName, "ci", textNode);
```
`ci` elements shouldn't contain units?

kerimoyle commented:
I can't add comments to some parts of the review pane:
- src/validator.cpp lines 1037, 1048: The description is odd, seems to be missing something in the middle or else why the gap after element?  `err->setDescription("Math has a '" + childNode->name() + "' element" + " that is not a supported MathML element.");`

- I'm confused about whether or not units are allowed/required/forbidden on `ci` elements.  A discussion ages ago with (I think, @agarny or @MichaelClerx?) said that we didn't allow units to be specified on `ci` elements as this was then an over-definition, as the variables had to have their own units anyway?

hsorby commented:
`Math in component 'membrane' contains 'time' as a bvar ci element but it is already a variable name.` messages like this may be a hang over from an older spec.  I think this should not be an error but I don't think the spec. is clear on this.

hsorby commented:
> * src/validator.cpp lines 1037, 1048: 

Message is ok I think just needs a little tidying.

> * I'm confused about whether or not units are allowed/required/forbidden on `ci` elements.  

I'm going with 'not defined by the spec. ' until proven otherwise.

hsorby commented:
`W3C MathML DTD error: No declaration for attribute cellml of element math.` . this might indicate that the math is not being cleaned properly, something to look into further.

kerimoyle commented:
> . I'm going with 'not defined by the spec. ' until proven otherwise.

You're right, it's not spelled out in the spec, except to say that we don't permit other unexpected attributes inside tags, so why would we here ...?  It *is* spelled out as required on the `cn` tags though.  There is a discussion in the specification document around this already?

kerimoyle commented:
> Math in component 'membrane' contains 'time' as a bvar ci element but it is already a variable name. messages like this may be a hang over from an older spec. I think this should not be an error but I don't think the spec. is clear on this.

I think @agarny said earlier that it shouldn't be an error (have searched for comment but can't find it ...).  I don't think it can be an error, as we need to define the variables outside the mathml, so however the variable is used inside it (as `<ci>` or `<bvar>`) shouldn't matter?

RE units in <ci> elements, see discussion [here](https://docs.google.com/document/d/1MleZJF6DYp4FO4vxslAHInrRFMqrSpARa1s_dPhwYU0/edit?disco=AAAACq5-iuI): 

> so ci elements cannot have units? or just don't have to?

> That's right, they can't have them in CellML 1 or 2.0, and units are just defined in the variable.
> So the ci uses the unit that the variable has in the component that the math is defined in

hsorby commented:
Yes I have made changes somewhere that I have temporarily misplaced that will only deal with cellml units on cn elements and require all ci elements to reference a variable in the component.  Once I where I put it that is.

hsorby committed:
Merge branch 'develop' into maths_validation_i424

hsorby committed:
Initialise and reset the xml parser when used.

hsorby committed:
Only worry about cellml:units on cn elemetns in the math, let the MathML DTD worry about anything else.

hsorby committed:
Add test on libXml2 to track differences in dealing with namespaces.

hsorby committed:
Move debug code out of utilities.h into own header debug.h.

hsorby committed:
Remove try_run test for libXml2 again after finding bug in code.

hsorby committed:
Add debug.h to list of GIT_HEADER_FILES in src/CMakeLists.txt.

hsorby committed:
Add basic tests to highlight bug in libXml2 namespace removal.

hsorby committed:
Fix bug when removing namespace.  Clearing subtrees use of the namespace before removal.

hsorby committed:
Fix clang format errors.

hsorby commented:
All fixed up and ready to review now.

hsorby committed:
Merge branch 'develop' into maths_validation_i424

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -608,20 +605,67 @@ TEST(Validator, invalidMathMLVariables)
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Model does not have a valid name attribute.",
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Component does not have a valid name attribute.",
```
Would it be practical to start reporting components with relationship to their parents now that we can do that easily?  It would make it a lot easier to locate the source of the error ... eg: "Component inside parent component "somethingHere" does not have a valid name" etc?

hsorby replied:
No, the error itself holds a reference to the component that has the error.

kerimoyle replied:
.. but the user doesn't get to see it unless they know to go hunting?  If you're interacting with libCellML only though a basic C++ file then this kind of error does nothing to help you.

nickerso replied:
The API docs should be sufficient to tell the user where to go looking for this information. Which we know is not the case currently and part of #334, so I think can resolve this in terms of this PR.

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -608,20 +605,67 @@ TEST(Validator, invalidMathMLVariables)
         "        <plus/>\n"
         "        <bvar>\n"
-        "          <ci cellml:units=\"dimensionless\" cellml:value=\"zero\">new_bvar</ci>\n"
+        "          <ci cellml:units=\"dimensionless\">new_bvar</ci>\n"
```
Also report on ci elements which have units?

hsorby replied:
No we are not doing that the spec. only mentions `cellml:units` on `cn` elements so letting the MathML DTD complain about units on `ci` elements.

kerimoyle replied:
Now looking below I understand this even less than before ... in line 700 are we saying that W3C needs a declaration of units for ci elements and that it's not there (but it is) or that it's not allowed? I am confused ...

nickerso replied:
Looks like this confusion has been resolved, right? We're only cleaning the MathML strings of what we know should be there, anything else may be picked up by validating against the W3C MathML DTD.

kerimoyle commented on tests/validator/validator.cpp:
```diff
@@ -647,16 +691,14 @@ TEST(Validator, invalidMathMLCiAndCnElementsWithCellMLUnits)
         "Math cn element with the value '2.0' does not have a valid cellml:units attribute.",
-        "W3C MathML DTD error: Namespace prefix cellml for value on ci is not defined.",
-        "No declaration for attribute cellml:value of element ci.",
+        "W3C MathML DTD error: No declaration for attribute units of element ci.",
```
Error message is ambiguous ... not sure whether it's saying we must or must not have units on ci

hsorby replied:
It's a message from libXml2 when using the MathML DTD we don't have any control over what is says.

kerimoyle reviewed (commented):
Looking good so far, but I'd like to make sure that the error messages returned are useful to the user.  This could perhaps be addressed en masse in a separate PR, but there are a couple that should be done here first.

hsorby committed:
Merge in changes from prime/develop fixing merge conflicts.

hsorby commented:
Ready for review.

kerimoyle commented:
Earlier issues from my review are still there ... Have we resolved where the units thing for <ci> tags sits yet?

hsorby commented:
I thought we had, we can only concern ourselves with `cellml:units` as an attribute on `cn` MathML elements.  Leaving others alone and letting the MathML DTD report errors on malformed MathML.

hsorby committed:
Merge branch 'develop' into maths_validation_i424

kerimoyle reviewed (dismissed):
Approving these so we can keep working, but would like to create/extend the issue around error messages as many of our current ones (including ones here) aren't as helpful as they could be.

hsorby committed:
Merge in changes from prime/develop fixing merge conflicts.

hsorby committed:
Merge branch 'maths_validation_i424' of github.com:hsorby/libcellml into maths_validation_i424

hsorby commented:
Ready for re-review.

nickerso commented on src/debug.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace libcellml {
+
+struct dbg
```
can this be consolidated with the `debug` that has gone into the test utilities? or this is really something different? (Separate issue, not something that I think needs to hold this one up!)

hsorby replied:
No it can't be consolidated with the `debug()` in test utilities.  I find it kind of useful but still has the same problem of leaving `#include <iostream>` around.

agarny replied:
At least, it would be nice to have `dbg` renamed to `Debug`...

nickerso commented on src/validator.cpp:
```diff
@@ -828,27 +799,18 @@ void Validator::ValidatorImpl::validateMath(const std::string &input, const Comp
+    validateAndCleanMathCiCnNodes(node, component, variableNames);
+
+    // Remove the cellml namespace definition.
+    if (mathNode->hasNamespaceDefinition(CELLML_2_0_NS)) {
```
why are we removing the CellML namespace?

hsorby replied:
Because this is about to be passed to the MathML DTD and if we have a stand alone math document it is likely to have a `cellml` namespace definition here.

nickerso committed:
Fix typo in comment

nickerso committed:
Adding additional test for MathML namespaces

Testing the behaviour when MathML namespace defined outside the `<math>` block.

nickerso reviewed (dismissed):
Couple of outstanding comments but nothing that should hold this one up. Assuming the CI comes back with a green tick I am happy with this one.

hsorby committed:
Merge branch 'develop' into maths_validation_i424

hsorby committed:
Merge branch 'maths_validation_i424' of github.com:hsorby/libcellml into maths_validation_i424

hsorby committed:
Move xml utility functions out of parser.cpp and into their own xmlutils compilation unit.

hsorby committed:
Rename dbg struct to Debug.

agarny commented on cmake/test_libxml2.cpp:
```diff
@@ -0,0 +1,32 @@
+  xmlCleanupParser();
+  xmlCleanupGlobals();	
+}
+
```
Am I missing something or is this file not used anywhere?

hsorby committed:
Remove test_libxml2.cpp file, this file is no longer used.

agarny committed:
Some minor cleaning up of debug.h.

agarny committed:
Added our copyright statement to xmlutils.cpp.

agarny committed:
Removed trailing spaces.

agarny committed:
Minor improvements to the code.

E.g.

```
bool a = false;
if (test) {
    a = true;
}
return a;
```

was replaced with:

```
if (test) {
    return true;
}
return false;
```

agarny committed:
Simplified the return of some strings.

agarny committed:
Make Clang-Tidy happy.

agarny commented on tests/parser/parser.cpp:
```diff
@@ -1858,6 +1858,110 @@ TEST(Parser, xmlComments)
+    EXPECT_EQ(size_t(0), parser.errorCount());
+}
+
+TEST(Parser, mathWithMathmlNamespaceOnModel)
```
Nice test!

agarny commented on tests/parser/parser.cpp:
```diff
@@ -1858,6 +1858,110 @@ TEST(Parser, xmlComments)
+    EXPECT_EQ(size_t(0), parser.errorCount());
+}
+
+TEST(Parser, mathWithNamespacesDefinedOnTheMathNodeNonStandardPrefix)
```
What was the intention here? The test name gives the impression that we wanted to test a non-standard prefix for the CellML namespace, which is the case here, but it's not definied on the math node (as claimed in the name of the test), but on the model (as can be seen in the code). So, which one should it be? (I don't think it matters since the most important, here, is probably to test the non-standard prefix?)

hsorby replied:
I think the test should be renamed to `mathWithNonStandardCellMLPrefix`.

agarny replied:
Sounds good to me. Done in commit d4327175.

agarny committed:
Renamed the mathWithNamespacesDefinedOnTheMathNodeNonStandardPrefix parser test...

... to mathWithNonStandardCellMLPrefix, which reflects more acccurately what it does.

agarny committed:
Some minor cleaning up.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -608,20 +605,67 @@ TEST(Validator, invalidMathMLVariables)
+TEST(Validator, invalidMathmlCellMLNsOnNode)
+{
+    const std::string math =
+        "<math  xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns=\"http://www.w3.org/1998/Math/MathML\"><apply><cellml:bvar><ci cellml:units=\"dimensionless\">B</ci></cellml:bvar></apply></math>";
```
I wish this test was based on a physical CellML file rather than inline code like here, which makes it more difficult to assess/review/understand/etc.

Comment also valid elsewhere.

nickerso replied:
whereas I like to see them inline to make it easy to see everything in one place :)

At least for these small test models.

agarny replied:
Sure, in this particular case, it was more about the lack of indentation and new lines. When it's some (short) CellML code that is indented and has new lines, then it's fine.

kerimoyle replied:
I much prefer cellml inline but with indents for code that's not a real model.  This means that it can be commented so we can follow the tests' purposes more easily, and saves hunting around for files when they only contain a couple of lines.  Indents are needed though ... :)

agarny committed:
Deleted [libCellML]/default.profraw.

This file is an LLVM Code Coverage related file, and has nothing to in our repository.

agarny committed:
Get our LLVM coverage back to 100%.

agarny reviewed (approved):
Looks good to me.
